### PR TITLE
Zero-out the prompt loss

### PR DESCRIPTION
**Pull request by l-k-11235**
This PR allows to zero-out the prompt loss when finetuning an LLM on a specific task.
This mechanism is mentioned in several papers, including [Self-Instruct: Aligning Language Models with Self-Generated Instructions](https://arxiv.org/abs/2212.10560)  and the Llama2 paper.

To enable it, the `zero_out_prompt_loss` config option must be set to true and the `insert_mask_before_placeholder` on-the-fly transform must be applied before the tokenization transform, eg:
```yaml
transforms: [insert_mask_before_placeholder, sentencepiece, filtertoolong]
zero_out_prompt_loss: true
```
This will lead to a higher accuracy.

### CODE REVIEW

**Comment by vince62s** [comment]
Many thanks for this.
If I read correctly it requires a specific preparation for the data and insert the MASK_BEFORE token between the prompt and the answer. What if instead we require to have the data prepared as SRC=prompts and TGT=answers. Wouldn't it be easier ?

File: onmt/utils/loss.py [hunk 1]
```diff
@@ -261,6 +261,40 @@ def _bottle(self, _v):
...
+        For finetuning on specific tasks.
+        The end of the prompt must be indicated by `the DefaultTokens.MASK_BEFORE`
+            placeholder.
+        The masks are supposed to be properly handled by the loss criterion
+            (e.g. nn.CrossEntropyLoss ).
+
+        Args:
+            batch: The current batch.
+        """
+        mask_before_idx = self.vocab["<blank>"]
```
Comment (francoishernandez) [inline onmt/utils/loss.py:277]: Not sure we want this to be hardcoded.

**Review by francoishernandez — COMMENTED** [review]
Looks good! Some tiny comments to address before we can merge.

**Comment by vince62s** [comment]
merging
